### PR TITLE
JAMES-2470 Reduce Cassandra table creation

### PR DESCRIPTION
**Pull request by chibenwa**
This is a long operation, and enforcing reuse leads to significant build time gains.

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[00307ed6bfd367d50d77b3d8d27d613a17128251] [INFO] --- maven-checkstyle-plugin:2.17:check (default) @ apache-james-backends-cassandra ---
[00307ed6bfd367d50d77b3d8d27d613a17128251] [INFO] There is 1 error reported by Checkstyle 8.5 with checkstyle.xml ruleset.
[00307ed6bfd367d50d77b3d8d27d613a17128251] [ERROR] src/main/java/org/apache/james/backends/cassandra/init/CassandraTableManager.java:[25,8] (imports) UnusedImports: Unused import - javax.inject.Inject.
```

**Comment by rouazana** [comment]
there are only fixups in this PR?

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/CassandraTableManager.java [hunk 1]
```diff
@@ -41,11 +51,25 @@ public CassandraTableManager ensureAllTables() {
...
+                .stream()
+                .map(CassandraTable::getName)
+                .map(name -> truncate(executor, name)))
+            .join();
+    }
+
+    private CompletableFuture<?> truncate(CassandraAsyncExecutor executor, String name) {
+        return executor.execute(
+            QueryBuilder.select().from(name).limit(1))
+            .thenCompose(resultSet -> truncateIfNeeded(executor, name, resultSet));
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/CassandraTableManager.java]: are you sure this "read before write" has a positive impact ?

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterWithKeyspaceCreatedFactory.java [hunk 2]
```diff
@@ -54,16 +54,11 @@ public Configuration replicationFactor(int replicationFactor) {
         }
         
         public Cluster clusterWithInitializedKeyspace() {
-            if (isKeyspacePresent(cluster, keyspace)) {
-                createKeyspace(cluster, keyspace, replicationFactor, durableWrites);
-            }
+            createKeyspace(cluster, keyspace, replicationFactor, durableWrites);
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterWithKeyspaceCreatedFactory.java]: how this change is related to this PR ?

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraCluster.java [hunk 3]
```diff
@@ -67,15 +57,19 @@ private CassandraCluster(CassandraModule module, @Named("cassandraHost") String
...
-            keyspace = new RandomStringGenerator.Builder().withinRange('a', 'z').build().generate(10);
-            clusterConfiguration = ClusterConfiguration.builder()
-                .host(Host.from(host, port))
-                .keyspace(keyspace)
-                .replicationFactor(1)
-                .maxRetry(10)
-                .minDelay(5000)
-                .build();
-            session = new FunctionRunnerWithRetry(MAX_RETRY).executeAndRetrieveObject(CassandraCluster.this::tryInitializeSession);
+            session = new SessionWithInitializedTablesFactory(
```
Comment (mbaechler) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraCluster.java]: How it is related to this PR ?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java [hunk 4]
```diff
@@ -17,33 +17,24 @@
...
-import org.apache.james.util.Host;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
 
-import com.google.inject.AbstractModule;
+import org.apache.james.backends.cassandra.init.CassandraTableManager;
+import org.apache.james.utils.GuiceProbe;
 
-public class CassandraJmapServerModule extends AbstractModule {
+public class CassandraCleanupProbe implements GuiceProbe {
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java]: why it needs to be a Probe ?

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/DockerCassandraRule.java [hunk 5]
```diff
@@ -43,15 +46,17 @@ public void await() {
 
     @Override
     public Module getModule() {
-        String keyspace = new RandomStringGenerator.Builder().withinRange('a', 'z').build().generate(12);
-        return (binder) -> binder.bind(ClusterConfiguration.class)
+        return Modules.combine((binder) -> binder.bind(ClusterConfiguration.class)
             .toInstance(ClusterConfiguration.builder()
                 .host(cassandraContainer.getHost())
-                .keyspace(keyspace)
+                .keyspace("testing")
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/DockerCassandraRule.java]: is it a duplicated String value ?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java [hunk 6]
```diff
@@ -17,33 +17,24 @@
...
     }
 
-    @Override
-    protected void configure() {
-        install(new CassandraTestModule(cassandraHost));
-        install(new TestElasticSearchModule(embeddedElasticSearch));
-        install(new TestJMAPServerModule(LIMIT_TO_3_MESSAGES));
-
-        install(binder -> binder.bind(TextExtractor.class).to(DefaultTextExtractor.class));
+    @PreDestroy
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java]: how are you handling the call ordering ? What if there's a lifecycle depending on a cassandra table ? what if TableManager is already "closed" ?

File: server/data/data-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryUrlStoreTest.java [hunk 7]
```diff
@@ -23,6 +23,6 @@
 import org.junit.jupiter.api.extension.ExtendWith;
 
 @ExtendWith(CassandraMailRepositoryUrlStoreExtension.class)
-public class CassandraMailRepositoryUrlStoreTest implements MailRepositoryUrlStoreContract {
+class CassandraMailRepositoryUrlStoreTest implements MailRepositoryUrlStoreContract {
```
Comment (mbaechler) [inline server/data/data-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryUrlStoreTest.java]: not related

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CassandraStepdefs.java [hunk 8]
```diff
@@ -70,9 +78,15 @@ public void init() throws Exception {
             .build();
 
         mainStepdefs.jmapServer = new GuiceJamesServer(configuration)
-                .combineWith(ALL_BUT_JMX_CASSANDRA_MODULE)
-                .overrideWith(new CassandraJmapServerModule(embeddedElasticSearch, cassandraServer.getHost()))
-                .overrideWith((binder) -> binder.bind(PersistenceAdapter.class).to(MemoryPersistenceAdapter.class));
+            .combineWith(ALL_BUT_JMX_CASSANDRA_MODULE)
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CassandraStepdefs.java]: How are these changes related to the PR ?

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CucumberCassandraSingleton.java [hunk 9]
```diff
@@ -18,7 +18,7 @@
  ****************************************************************/
 package org.apache.james.jmap.cassandra.cucumber;
 
-import org.apache.james.backends.cassandra.DockerCassandraRule;
+import org.apache.james.DockerCassandraRule;
 
 public class CucumberCassandraSingleton {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CucumberCassandraSingleton.java]: do we still need this class now ?

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraCluster.java [hunk 10]
```diff
@@ -67,15 +57,19 @@ private CassandraCluster(CassandraModule module, @Named("cassandraHost") String
...
                 .build();
-            keyspace = new RandomStringGenerator.Builder().withinRange('a', 'z').build().generate(10);
-            clusterConfiguration = ClusterConfiguration.builder()
-                .host(Host.from(host, port))
-                .keyspace(keyspace)
-                .replicationFactor(1)
-                .maxRetry(10)
-                .minDelay(5000)
-                .build();
-            session = new FunctionRunnerWithRetry(MAX_RETRY).executeAndRetrieveObject(CassandraCluster.this::tryInitializeSession);
```
Comment (trantienduchn) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraCluster.java]: Why did you remove retry?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java (hunk 4 above)
Comment (trantienduchn) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java]: are you sure to place this class in `src/main` instead of `src/test` folder?

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterWithKeyspaceCreatedFactory.java (hunk 2 above)
Comment (chibenwa) [reply backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterWithKeyspaceCreatedFactory.java]: The boot time optimization of looking if the keyspace exist before creating it is pointless now that we recycle keyspace.

Furthermore a "IF NOT EXISTS" is used while creating the keyspace.

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraCluster.java (hunk 3 above)
Comment (chibenwa) [reply backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraCluster.java]: Well, once I'm at it, CassandraCluster don't need to await anything, so, you know.... I consider usefull to stop lying to people.

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java (hunk 4 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java]: To force its instanciation via guice, so that it gets run upon destroy.

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/DockerCassandraRule.java (hunk 5 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/DockerCassandraRule.java]: God damn no! I'm not that kind of criminal! It's two indepandant definitions of keyspaces that happens to have the same name, that's all...

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java (hunk 6 above)
Comment (mbaechler) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java]: It doesn't in my understanding (or maybe just by accident). You use multibind to register your Probe, it uses a Set and AFAIK doesn't guarantee any ordering. How do you know it will respect the order you want ? I may I missed something ....

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CassandraStepdefs.java (hunk 8 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CassandraStepdefs.java]: `CassandraJmapServerModule` was used only here and **messed things up**.

I remove it as it bringed no value. And rather write the needed overrides here.

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CassandraStepdefs.java [hunk 11]
```diff
@@ -70,9 +78,15 @@ public void init() throws Exception {
...
+            .overrideWith(binder -> binder.bind(TextExtractor.class).to(PDFTextExtractor.class))
+            .overrideWith(new TestJMAPServerModule(10))
+            .overrideWith(new TestESMetricReporterModule())
+            .overrideWith(new TestElasticSearchModule(embeddedElasticSearch))
+            .overrideWith(cassandraServer.getModule())
+            .overrideWith(binder -> binder.bind(TextExtractor.class).to(DefaultTextExtractor.class))
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CassandraStepdefs.java]: we bind two TextExtractor

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java (hunk 6 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java]: The probe is staged after init, because the GuiceProbeProvider is recovered after performing init.

With the stager, it's not the inject which counts but the actual init <3

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java (hunk 4 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java]: Well, we could desine a new kind of registries, and affect **GuiceJamesServer** method with it specially for this purpose. But impacting important classes like this with only test concerns seems bad to me.

Or I could register a configuration perform that actually triggers the inject, but it seems even morebloated with me, and might break what I describe in https://github.com/linagora/james-project/pull/1550#discussion_r203646979

Comment (mbaechler) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java]: > But impacting important classes like this with only test concerns seems bad to me.

It's always like that one aim of writing tests is to write production code that is more flexible, better designed, etc. In this case, we lack things in James to manage lifecycle, I don't see any good reason to not fix that issue, it will just make James a better software. Our users would benefit from that because they may use James in context we don't think of today.

TL;DR
I really think we should not trick the code but do design in these kind of cases.

Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/server/CassandraCleanupProbe.java]: Ok then let's do a cleanup registry.

**Comment by chibenwa** [comment]
```
[ad20e07fec2570082a64ee0d6b467f45281e21a9] Fetching upstream changes from https://github.com/chibenwa/james-project.git
[ad20e07fec2570082a64ee0d6b467f45281e21a9]  > git -c core.askpass=true fetch --tags --progress https://github.com/chibenwa/james-project.git +refs/heads/*:refs/remotes/origin/*
[ad20e07fec2570082a64ee0d6b467f45281e21a9] ERROR: Error fetching remote repo 'origin'
[ad20e07fec2570082a64ee0d6b467f45281e21a9] hudson.plugins.git.GitException: Failed to fetch from https://github.com/chibenwa/james-project.git
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:735)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:983)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1016)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:109)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:83)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:73)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1$1.call(AbstractSynchronousNonBlockingStepExecution.java:49)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.security.ACL.impersonate(ACL.java:213)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1.run(AbstractSynchronousNonBlockingStepExecution.java:47)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.lang.Thread.run(Thread.java:745)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] Caused by: hudson.plugins.git.GitException: Command "git -c core.askpass=true fetch --tags --progress https://github.com/chibenwa/james-project.git +refs/heads/*:refs/remotes/origin/*" returned status code 128:
[ad20e07fec2570082a64ee0d6b467f45281e21a9] stdout: 
[ad20e07fec2570082a64ee0d6b467f45281e21a9] stderr: fatal: unable to access 'https://github.com/chibenwa/james-project.git/': gnutls_handshake() failed: The TLS connection was non-properly terminated.
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1591)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1379)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:86)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:324)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.remoting.UserRequest.perform(UserRequest.java:181)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.remoting.UserRequest.perform(UserRequest.java:52)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.remoting.Request$2.run(Request.java:336)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.remoting.Engine$1$1.run(Engine.java:98)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.lang.Thread.run(Thread.java:748)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at ......remote call to obiwan(Native Method)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1416)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.remoting.UserResponse.retrieve(UserRequest.java:220)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.remoting.Channel.call(Channel.java:781)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at sun.reflect.GeneratedMethodAccessor514.invoke(Unknown Source)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at java.lang.reflect.Method.invoke(Method.java:498)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at com.sun.proxy.$Proxy76.execute(Unknown Source)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:733)
[ad20e07fec2570082a64ee0d6b467f45281e21a9] 	... 13 more
[ad20e07fec2570082a64ee0d6b467f45281e21a9] Running: Shell Script
```

What?

test this please

File: server/container/guice/guice-common/src/main/java/org/apache/james/GuiceJamesServer.java [hunk 12]
```diff
@@ -80,7 +81,7 @@ public void start() throws Exception {
 
     public void stop() {
         if (preDestroy != null) {
-            cleanupTasksPerformer.clean();
+            Optional.ofNullable(cleanupTasksPerformer).ifPresent(CleanupTasksPerformer::clean);
```
Comment (mbaechler) [inline server/container/guice/guice-common/src/main/java/org/apache/james/GuiceJamesServer.java]: I think it should not be in `if (preDestroy != null)` block

Comment (chibenwa) [reply server/container/guice/guice-common/src/main/java/org/apache/james/GuiceJamesServer.java]: true. I fixed this.
